site/source/docs/compiling/WebAssembly.rst

```diff
@@ -23,7 +23,7 @@ WebAssembly is emitted by default, without the need for any special flags.
   emcc [..args..] -s WASM=0
 
 .. note:: Emscripten's WebAssembly support depends on `Binaryen <https://github.com/WebAssembly/binaryen>`_, which is provided by the emsdk (if you don't use the emsdk, you need to build it and set it up in your ``.emscripten`` file).
-.. note:: The ``WASM``, ``BINARYEN*``, etc. options only matter when compiling to your final executable. In other words, the same .o files are used for both asm.js and WebAssembly. Only when linking them and compiling to asm.js or WebAssembly do you need to specify WebAssembly if you want that. That means that it is easy to build your project to both asm.js and WebAssembly. (The only exception is the wasm backend's ``WASM_OBJECT_FILES`` flag which determines whether object files contain wasm or LLVM IR.)
+.. note:: Deciding to compile to wasm or JS can be done at the linking stage: it doesn't affect the object files.
```

Comment:
Does it happens at the linking stage? I think if `WASM_OBJECT_FILES = 1`, for the default mode, compiling to WebAssembly generates wasm object files (.o) and lld links them, but compiling to JS (fastcomp) generates LLVM bitcode files (which is .bc, even though they are named .o in emscripten pipeline), and they are merged by llvm-link..?

Reply:
It's true that the object files are different between all those modes, but the decision to emit wasm or asm.js/js happens at link time: with fastcomp, we just do the asm.js backend and stop before asm2wasm, leaving asm.js, and with the wasm backend for js we run wasm2js.

In other words it's actually pretty simple, the decision to do js or wasm is purely at link time, no matter the backend, which I didn't realize until now.

Reply:
Sorry I don't understand. fastcomp does not even emit wasm object files to begin with, no? Doesn't fastcomp backend emit .js?

Reply:
fastcomp emits bitcode in object files, correct. but fastcomp can be used to emit either asm.js or wasm, and that decision is made at link time (we get asm.js and then either leave it as such, or compile that to wasm using asm2wasm).